File: SentAnalysis.java
package bigdata.sparky;

import static org.apache.spark.sql.functions.col;

import java.sql.Time;
import java.util.Arrays;
import java.util.List;

import org.apache.log4j.Level;
import org.apache.log4j.Logger;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.mllib.regression.LabeledPoint;
import org.apache.spark.mllib.feature.HashingTF;
import org.apache.spark.mllib.tree.GradientBoostedTrees;
import org.apache.spark.mllib.tree.configuration.BoostingStrategy;
import org.apache.spark.mllib.tree.configuration.Strategy;
import org.apache.spark.mllib.tree.model.GradientBoostedTreesModel;
import org.apache.spark.mllib.tree.model.Predict;
import org.apache.spark.mllib.linalg.Vector;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.SparkSession.Builder;
import org.apache.spark.sql.types.StructType;

import scala.Tuple2;
import shapeless.labelled;


public class SentAnalysis {
	
	private static final String topic = "pizza";
	
	/*
	 * Some variables using for the evaluating of the model
	 */
	static long happyTotal = 0;
	static long sadTotal = 0;
	static long happyCorrect = 0;
	static long sadCorrect = 0;
	
	/*
	 * The Schema of tweets' text
	 * We are going to use only Text
	 * 
	 * 		tweetText(tweet_id,user_id,text)
	 * 
	 */
	static StructType schemaText = new StructType()
			.add("tweet_id", "long")
			.add("user_id", "long")
			.add("text", "String")
			.add("geo_lat", "long")
			.add("geo_long", "long")
			.add("place_full_name", "String")
			.add("place_id", "long");

	/*
	 * 
	 * 
	 * @author Raffaele Perini, Giovanni Rafaèl Vuolo
	 * 
	 * 
	 * 
	 */
	public static void main(String[] args) {

		// SparkSession
		Builder builder = new Builder().appName("SparkSQL Examples");
		builder.master("local");


		SparkSession spark = builder.getOrCreate();


		Logger.getLogger("org").setLevel(Level.OFF);
		Logger.getLogger("akka").setLevel(Level.OFF);


		/*
		 * 	Regarding the tweet's texts all we need is tweet_id, user_id, text
		 *  So we do a select on the tweet table
		 *  then we print the schema	
		 */
		Dataset<Row> text = spark
				.read()
				.option("mode", "DROPMALFORMED")
				.schema(schemaText)
				.csv("/home/jave/Documents/UNIMASTER/bigData/data_02/tweet_text_02.csv");

		text = text.select(col("text"));

		text.printSchema();


		/*
		 * @FROMRAWDATATODATAFRAME
		 * Let's start with clean the record, 
		 * We want to remove any tweet that doesn’t contain “happy” or “sad”. Why? 'coz 
		 * we can easily make the assumption that if a tweet contains happy it's an happy 
		 * tweet, and the same for sad.
		 * What about the "I am not happy" case???? --> Not handle
		 */ 

		//text.select("text").createOrReplaceTempView("text");
		long analysedTweetsNumber = text.count();
		System.out.println("There are : " + analysedTweetsNumber + " tweets");

		/*
		 * Keep only the tweet that contains happy or sad
		 */
		Dataset<Row> happyTweet = text.filter(col("text").contains("happy"));
		long counterOfHappyTweets = happyTweet.count();
		happyTweet.show();
		Dataset<Row> sadTweet = text.filter(col("text").contains("sad"));
		long counterOfSadTweets = sadTweet.count();
		sadTweet.show();
		System.out.println("#happy: " + counterOfHappyTweets + "\n#sad: " + counterOfSadTweets);
		
		/*
		 * We keep an equal number sad and happy tweets in order prevent bias in the model
		 * then we create a new view using an an equal number of happy and sad tweets
		 */
		long minNumber = Math.min(counterOfHappyTweets, counterOfSadTweets);
		System.out.println(minNumber);
		Dataset<Row> tmp1 = happyTweet.limit((int) minNumber).union(sadTweet.limit((int) minNumber));	//LOOK FOR UNIONALL
		tmp1.show();

		/*
		 * @LABELINGOFDATA
		 * We remove the word happy and sad (or similar) from the tweet in order to 
		 * infer the happiness or the sadness only via the other words, and also 
		 * we label each tweet as 1 if happy or 0 if sad
		 */

		JavaRDD<Row> textRdd = tmp1.javaRDD();

		JavaRDD<Tuple2<Integer,String[]>> statusAndSplit = textRdd.map(
				(Function<Row,Tuple2<Integer,String[]>>)s -> {
					try {
						String txt = s.getString(0).toLowerCase();
						int isHappy = 0;
						if(txt.contains("happy")) {
							isHappy = 1;
						}else if(txt.contains("sad")) {
							isHappy = 0;
						}
						txt = txt.replaceAll("happy", "");
						txt = txt.replaceAll("sad", "");
						Tuple2<Integer,String[]> ret = new Tuple2<>(isHappy, txt.split(" "));
						return ret;

					}catch(Exception e){
						System.out.println(e);
					};
					return null;
				});

		long exceptions = textRdd.count() - statusAndSplit.count();
		System.out.println("Number of records now:" + statusAndSplit.count());
		System.out.println("Number of exceptions: " + exceptions);

		for(int i = 0; i < 5; i++) {
			System.out.println(statusAndSplit.take(5).get(i));
		}

		/*
		 * @DATATRANSFORMATION
		 * We use a Gradient Boosting algorithm that excepts an array of fixed lenght of number
		 * we hash each word into an fixed-lenght array
		 * we get an array that represents the count of ecah word in the tweet
		 * we use Bag-Of-Words 
		 * for implement bow we use hashingTF
		 * we use an array of 3000 that seems it is enough
		 * since 3000 < #ofwords it is possibile that two or more words 
		 */

		HashingTF hashingTF = new HashingTF(3000);

		JavaRDD<LabeledPoint> textRdd2 = 
				statusAndSplit
				.map((Function<Tuple2<Integer,String[]>,Tuple2<Integer,Vector>>) t -> {
					return new Tuple2<Integer,Vector>(
							t._1, 
							hashingTF.transform(Arrays.asList(t._2)));
				})
				.map((Function<Tuple2<Integer,Vector>,LabeledPoint>)t -> {
					return new LabeledPoint(t._1.doubleValue(), t._2);
				});

		for(int i = 0; i < 10; i++) {
			System.out.println(textRdd2.take(10).get(i));
		}
		
		/*
		 * @SPLITOFSETS
		 * 30% for the validation set
		 * 
		 */
		
		
		// Split the data into training and test sets (30% held out for testing)
		JavaRDD<LabeledPoint>[] splits = textRdd2.randomSplit(new double[] {0.7, 0.3});
		JavaRDD<LabeledPoint> trainingData = splits[0];
		JavaRDD<LabeledPoint> validationData = splits[1];
		
		
		/*
		 * @BUILDTHEMODEL
		 * 30 passes over the trainingset
		 * 2 classes, happy or sad
		 * 6 depth of each tree. The higher it is, the higher probability of overfitting
		 * the lower it is, the simpler the model is
		 */
		
		BoostingStrategy bs = BoostingStrategy.defaultParams("Classification");
		bs.setNumIterations(30);
		bs.treeStrategy().setNumClasses(2);
		bs.treeStrategy().setMaxDepth(10);
		
		GradientBoostedTreesModel model = GradientBoostedTrees.train(trainingData, bs);
		
		/*
		 * @EVALUATETHEMODEL
		 * 
		 * 
		 */
		
		JavaRDD<Tuple2<Double,Double>> labelAndPredictTrain = trainingData.map(
				x -> {
					double prediction = model.predict(x.features());
					return new Tuple2<Double,Double>(x.label(), prediction);
				});
				
		JavaRDD<Tuple2<Double,Double>> labelAndPredictValid = validationData.map(
				x -> {
					double prediction = model.predict(x.features());
					return new Tuple2<Double,Double>(x.label(), prediction);
				});
		
		
		List<Tuple2<Double,Double>> results = labelAndPredictTrain.collect();
	
		results.forEach(
				x -> {
					if(x._1 == 1) {happyTotal++;}
					else if(x._1 == 0) {sadTotal++;}
					if(x._1 == 1 && x._2 == 1) {happyCorrect++;}
					else if(x._1 == 0 && x._2 == 0) {sadCorrect++;}
				});
		
		System.out.println("\n\n=========" );
		System.out.println(" TOPIC ----------------> " + topic  );
		System.out.println(" #Analysed Tweets -----> " + analysedTweetsNumber + "\n");
		
		
		System.out.println("=====TRAIN=====" );
		System.out.println("Total happys+sad: " + (happyTotal + sadTotal));

		System.out.println("=HAPPY=: " );
		System.out.println("Total happys: " + happyTotal );
		System.out.println("Total correct happys: " + happyCorrect );
		System.out.println("Percentage of correct happys : " + new Double(happyCorrect)/new Double(happyTotal) );
		
		System.out.println("=SAD=: " );
		System.out.println("Total sads: " + sadTotal );
		System.out.println("Total correct sads: " + sadCorrect );
		System.out.println("Percentage of correct sads : " + new Double(sadCorrect)/ new Double(sadTotal) );
		
		System.out.println("=ERROR=: " );
		long tmpEr = (happyTotal + sadTotal) - (happyCorrect + sadCorrect);
		double error = new Double(tmpEr) / new Double(happyTotal + sadTotal);
		System.out.println("Total error: " + tmpEr );
		System.out.println("Total error percentage: " + error );
		
		
		
		results = labelAndPredictValid.collect();
		
		results.forEach(
				x -> {
					if(x._1 == 1) {happyTotal++;}
					else if(x._1 == 0) {sadTotal++;}
					if(x._1 == 1 && x._2 == 1) {happyCorrect++;}
					else if(x._1 == 0 && x._2 == 0) {sadCorrect++;}
				});
		
		System.out.println("\n\n=====VALID=====" );
		System.out.println("Total happys+sad: " + (happyTotal + sadTotal));

		System.out.println("=HAPPY=: " );
		System.out.println("Total happys: " + happyTotal );
		System.out.println("Total correct happys: " + happyCorrect );
		System.out.println("Percentage of correct happys : " + new Double(happyCorrect)/new Double(happyTotal) );
		
		System.out.println("=SAD=: " );
		System.out.println("Total sads: " + sadTotal );
		System.out.println("Total correct sads: " + sadCorrect );
		System.out.println("Percentage of correct sads : " + new Double(sadCorrect)/ new Double(sadTotal) );
		
		System.out.println("=ERROR=: " );
		tmpEr = (happyTotal + sadTotal) - (happyCorrect + sadCorrect);
		error = new Double(tmpEr) / new Double(happyTotal + sadTotal);
		System.out.println("Total error: " + tmpEr );
		System.out.println("Total error percentage: " + error );
		
		System.out.println(new Time(System.currentTimeMillis()).toGMTString().toString());
		
		labelAndPredictTrain.saveAsTextFile("hdfs://localhost:9000/bigdata/project/train.csv");
		labelAndPredictValid.saveAsTextFile("hdfs://localhost:9000/bigdata/project/valid.csv");
	}
}

File: SentAnalysis.scala

import org.apache.log4j.{Level, Logger}
import org.apache.spark.mllib.feature.HashingTF
import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.mllib.regression.LabeledPoint
import org.apache.spark.mllib.tree.GradientBoostedTrees
import org.apache.spark.mllib.tree.configuration.BoostingStrategy
import org.apache.spark.sql.SparkSession

import scala.util.{Success, Try}



object RunSent{

  def main(args: Array[String]): Unit = {

    Logger.getLogger("org").setLevel(Level.OFF)
    Logger.getLogger("akka").setLevel(Level.OFF)


    val conf = new SparkConf().setAppName("mini-project2").setMaster("local")
    val sc = new SparkContext(conf)

    val spark = SparkSession
      .builder()
      .appName("mini-project2")
      .config("spark.some.config.option", "some-value")
      .getOrCreate()

    /*
		 * @FROMRAWDATATODATAFRAME
		 * We use a Twitter dataset, but we care only about the tweet text
		 */

    val tweetDF = spark.read.option("mode", "DROPMALFORMED").csv("tweets.csv")
    tweetDF.show()

    var messages = text.select("text")
    println("Total messages: " + messages.count())

    
    /*
		 * @LABELINGOFDATA
		 * I use an NLP <see other class>
		 */

    val tmp= SentimentTraining.trainingMe(messages).rdd
    val records = tmp.map(
      row =>{
        Try{
          val sentiment= row(1).toString.toLowerCase()
          var msgSanitized = sentiment.replace("max", "max") // placeholder
          var isHappy = 0
          if(sentiment.contains("positive")){
            isHappy = 1
            msgSanitized = sentiment.replace("positive", "1")
          }else if(sentiment.contains("negative")){
            msgSanitized = sentiment.replace("negative", "0")
            isHappy = 0
          }
          //Return a tuple
          (isHappy, row(0).toString.split(" ").toSeq)
        }
      }
    )

    println("RECORDS")
    records.take(20).foreach(x => println(x+"e"))


    //We use this syntax to filter out exceptions
    val exceptions = records.filter(_.isFailure)
    println("total records with exceptions: " + exceptions.count())

    exceptions.take(20).foreach(x => println(x+"e"))

    var labeledTweets = records.filter((_.isSuccess)).map(_.get)
    println("total records with successes: " + labeledTweets.count())

    labeledTweets.take(20).foreach(x => println(x+"l"))


    /*
		 * @DATATRANSFORMATION
		 * We use a Gradient Boosting algorithm that excepts an array of fixed lenght of number
		 * we hash the words into an fixed-lenght array
		 * we get an array that represents the count of ecah word in the tweet
		 * we use Bag-Of-Words
		 * for implement bow we use hashingTF
		 * we use an array of 3000 that seems it is enough
		 * since 3000 < #ofwords it is possible that two or more words
		 */
    val hashingTF = new HashingTF(2000)

    //Map the input strings to a tuple of labeled point + input text
    val input_labeled = (labeledTweets.map(
      t => (t._1, hashingTF.transform(t._2)))
      .map(x => new LabeledPoint((x._1).toDouble, x._2)))

    input_labeled.take(10).foreach(println)

  
    /*
		 * @SPLITOFSETS
		 * 30% for the validation set
		 *
		 */
    val splits = input_labeled.randomSplit(Array(0.7, 0.3))
    val (trainingData, validationData) = (splits(0), splits(1))


    /*
		 * @BUILDTHEMODEL
		 * 30 passes over the trainingset
		 * 2 classes, happy or sad
		 * 6 depth of each tree. The higher it is, the higher probability of overfitting
		 * the lower it is, the simpler the model is
		 */
    val boostingStrategy = BoostingStrategy.defaultParams("Classification")
    boostingStrategy.setNumIterations(25) //number of passes over our training data
    boostingStrategy.treeStrategy.setNumClasses(2) //We have two output classes: happy and sad
    boostingStrategy.treeStrategy.setMaxDepth(7)

    val model = GradientBoostedTrees.train(trainingData, boostingStrategy)

    /*
		 * @EVALUATETHEMODEL
		 *
		 * and compute error
		 */
    var labelAndPredsTrain = trainingData.map { point =>
      val prediction = model.predict(point.features)
      Tuple2(point.label, prediction)
    }

    var labelAndPredsValid = validationData.map { point =>
      val prediction = model.predict(point.features)
      Tuple2(point.label, prediction)
    }

    //Since Spark has done the heavy lifting already, lets pull the results back to the driver machine.
    //Calling collect() will bring the results to a single machine (the driver) and will convert it to a Scala array.

    //Start with the Training Set
    val results = labelAndPredsTrain.collect()

    var happyTotal = 0
    var unhappyTotal = 0
    var happyCorrect = 0
    var unhappyCorrect = 0

    val resultsTr = labelAndPredsTrain.collect().foreach(
      r => {
        if (r._1 == 1) {
          happyTotal += 1
        } else if (r._1 == 0) {
          unhappyTotal += 1
        }
        if (r._1 == 1 && r._2 ==1) {
          happyCorrect += 1
        } else if (r._1 == 0 && r._2 == 0) {
          unhappyCorrect += 1
        }
      }
    )
    println("Training Set:")
    println("    unhappy messages " + unhappyTotal)
    println("    happy messages: " + happyTotal)
    println("    happy % correct: " + happyCorrect.toDouble/happyTotal)
    println("    unhappy % correct: " + unhappyCorrect.toDouble/unhappyTotal)

    val testErrTr = labelAndPredsTrain.filter(r => r._1 != r._2).count.toDouble / trainingData.count()
    println("    test Error :" + testErrTr)


    happyTotal = 0
    unhappyTotal = 0
    happyCorrect = 0
    unhappyCorrect = 0
    val resultsVa = labelAndPredsValid.collect().foreach(
      r => {
        if (r._1 == 1) {
          happyTotal += 1
        } else if (r._1 == 0) {
          unhappyTotal += 1
        }
        if (r._1 == 1 && r._2 ==1) {
          happyCorrect += 1
        } else if (r._1 == 0 && r._2 == 0) {
          unhappyCorrect += 1
        }
      }
    )
    println("Validation Set ")
    println("    unhappy messages" + unhappyTotal)
    println("    happy messages: " + happyTotal)
    println("    happy % correct: " + happyCorrect.toDouble/happyTotal)
    println("    unhappy % correct: " + unhappyCorrect.toDouble/unhappyTotal)

    val testErrVa = labelAndPredsValid.filter(r => r._1 != r._2).count.toDouble / validationData.count()
    println("    test error: " + testErrVa)




  }

}
